_drafts/firmware-code-size-optimization-1.md

```diff
@@ -0,0 +1,302 @@
+more code space is always better.
+
+In this series of posts, we'll explore ways to save code space and ways not to
+do it. We will cover compiler options, coding style, logging, as well as
+desperate hacks when all you need is another 24 bytes.
+
+But first, let's talk about measuring code size.
+
+<!-- excerpt end -->
+
+### Why measure?
+
+Optimization is often counter intuitive. Don't take my word for it: this is one
+of the topics [Microsoft's Raymond
+Wong](https://devblogs.microsoft.com/oldnewthing/20041216-00/?p=36973) and [Unix
+programmers](https://homepage.cs.uri.edu/~thenry/resources/unix_art/ch12s02.html)
+agree on!
+
+So before you do anything, measure where your code size is going. You may be
+surprised!
+
+### Measuring overall code size
+
+The simplest way to measure code size is to inspect the different sections of
+your elf file. The ARM GCC toolchain comes with a handy utility to do just that:
+`arm-none-eabi-size`.
+
+Let's try running it on one of our ELF files:
+
+```terminal
+francois-mba:with-libc francois$ arm-none-eabi-size build/with-libc.elf
+   text    data     bss     dec     hex filename
+  10800     104    8272   19176    4ae8 build/with-libc.elf
+```
+
+So in this program we have:
+* 10800 bytes of `text`, which is code
+* 104 bytes of `data`, which is statically initalized memory
+* 8272 bytes of `bcc`, which is zero-initialized memory
+
+Unfortunately, the output of this tool is a little misleading. You may think
+that your program will occupy `10800` bytes of flash, but this is not the case.
+
+Indeed if we modify our linker script to specify that our flash size is 10800
+exactly:
+
+```diff
+diff --git a/with-libc/samd21g18a_flash.ld b/with-libc/samd21g18a_flash.ld
+index a07cc82b..f49e73e2 100644
+--- a/with-libc/samd21g18a_flash.ld
++++ b/with-libc/samd21g18a_flash.ld
+@@ -3,7 +3,7 @@ OUTPUT_ARCH(arm)
+ 
+ MEMORY
+ {
+-  rom      (rx)  : ORIGIN = 0x00000000, LENGTH = 0x00040000
++  rom      (rx)  : ORIGIN = 0x00000000, LENGTH = 10800
+   ram      (rwx) : ORIGIN = 0x20000000, LENGTH = 0x00008000
+ }
+```
+
+We end up overflowing our flash!
+
+```terminal
+francois-mba:with-libc francois$ make
+[...]
+/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld:
+build/with-libc.elf section `.data' will not fit in region `rom'
+/usr/local/Cellar/arm-none-eabi-gcc/8-2018-q4-major/gcc/bin/../lib/gcc/arm-none-eabi/8.2.1/../../../../arm-none-eabi/bin/ld:
+region `rom' overflowed by 104 bytes
+collect2: error: ld returned 1 exit status
+make: *** [build/with-libc.elf] Error 1
+```
+
+The 104 bytes overflow hints at the cause: we are overflowing our flash by
+the size of our `data` section. This is because initialization values for our
+initialized static variables are stored in flash as well.
+
+To avoid confusion, I prefer to use a small script to compute the sizes:
+
+```bash
+#!/bin/bash
```

Comment:
MAYBE: Might be neat to publish a python package that does this type of thing.

Reply:
This would be ~1 day of work